tools/tpm2_policysecret.c

```diff
@@ -0,0 +1,215 @@
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+#include <string.h>
+#include <tss2/tss2_sys.h>
+
+#include "tpm2_auth_util.h"
+#include "files.h"
+#include "log.h"
+#include "tpm2_options.h"
+#include "tpm2_policy.h"
+#include "tpm2_session.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct tpm2_policysecret_ctx tpm2_policysecret_ctx;
+struct tpm2_policysecret_ctx {
+    //File path for the session context data
+    const char *session_path;
+    //File path for storing the policy digest output
+    const char *out_policy_dgst_path;
+    //Specifying the TPM object handle-id or the file path of context blob
+    const char *context_arg;
+    tpm2_loaded_object context_object;
+    //Auth value of the auth object
+    char *auth_str;
+    //Auth value length
+    UINT16 auth_str_size;
+    struct {
+        UINT8 c : 1;
+        UINT8 p : 1;
+    } flags;
+    TPMS_AUTH_COMMAND session_data;
+};
+
+static tpm2_policysecret_ctx ctx = {
+    .session_data = TPMS_AUTH_COMMAND_INIT(TPM2_RS_PW),
+    .auth_str_size = UINT16_MAX,
+    .auth_str = NULL,
+};
+
+static bool on_option(char key, char *value) {
+
+    bool result = true;
+
+    switch (key) {
+    case 'o':
+        ctx.out_policy_dgst_path = value;
+        break;
+    case 'S':
+        ctx.session_path = value;
+        break;
+    case 'c':
+        ctx.context_arg = value;
+        ctx.flags.c = 1;
+        break;
+    }
+
+    return result;
+}
+
+bool on_arg (int argc, char **argv) {
+
+    if (argc > 1) {
+        LOG_ERR("Specify stdin with '-' ");
+        return false;
+    }
+
+    if (!argc) {
+        //empty auth
+        return true;
+    }
+
+    bool result;
+    if (!strcmp(argv[0], "-")) {
+        //auth data from stdin
+        result = files_load_bytes_from_buffer_or_file_or_stdin(NULL,NULL,
+                &ctx.auth_str_size, ctx.session_data.hmac.buffer);
+        if (!result) {
+            LOG_ERR("Failed stdin input for auth.");
```

Comment:
I think files_load_bytes_from_buffer_or_file_or_stdin() logs all errors. So just drop this or fix up the sentence:
```C
LOG_ERR("Failed to read auth value from stdin");
```